File: src/main/kotlin/com/rstudio/shinycannon/Detect.kt
package com.rstudio.shinycannon

import net.moznion.uribuildertiny.URIBuilderTiny
import org.apache.http.client.methods.HttpGet
import org.apache.logging.log4j.Logger
import kotlin.system.exitProcess

enum class ServerType(val typeName: String) {
    RSC("RStudio Server Connect"),
    SSP("Shiny Server or Shiny Server Pro"),
    SAI("shinyapps.io"),
    SHN("R/Shiny")
}

fun typeFromName(typeName: String): ServerType {
    return when (typeName) {
        "RStudio Server Connect" -> ServerType.RSC
        "Shiny Server or Shiny Server Pro" -> ServerType.SSP
        "shinyapps.io" -> ServerType.SAI
        "R/Shiny" -> ServerType.SHN
        else -> error("Unknown server type name in recording: ${typeName}")
    }
}

fun servedBy(appUrl: String, logger: Logger): ServerType {
    val url = URIBuilderTiny(appUrl)

    if (url.host.matches("^.*\\.shinyapps\\.io$".toRegex()))
        return ServerType.SAI

    val resp = slurp(HttpGet(appUrl))

    if (resp.headers.containsKey("SSP-XSRF")) {
        return ServerType.SSP
    } else if (resp.headers.containsKey("x-powered-by")) {
        val sspVals = setOf("Express", "Shiny Server", "Shiny Server Pro")
        if (sspVals.contains(resp.headers["x-powered-by"]))
            return ServerType.SSP
    } else if (resp.cookies.cookies.firstOrNull { it.name == "rscid" } != null) {
        return ServerType.RSC
    }

    val shinyJsNode = try {
        xpath(resp.body, "/html/head/script")
                .flatMap { List(it.attributes.length, { i -> it.attributes.item(i)}) }
                .filter { it.nodeName == "src" }
                .find { it.nodeValue.matches("^.*/shiny(\\.min)?\\.js$".toRegex()) }
    } catch (e: Exception) {
        null
    }

    if (shinyJsNode == null) {
        logger.error("Target URL ${appUrl} does not appear to be a Shiny application.")
        exitProcess(1)
    }

    return ServerType.SHN
}

File: src/detect.ts
/**
 * Server type auto-detection.
 *
 * Determines which server type is hosting a Shiny application by
 * inspecting the hostname and the HTTP response from a GET to the app URL.
 */

import { ServerType } from "./types.js"
import type { HttpClient } from "./http.js"

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/**
 * Returns true if `body` contains a script reference to shiny.js or
 * shiny.min.js (e.g. `<script src="shared/shiny.min.js">`).
 */
export function hasShinyJs(body: string): boolean {
  return /\/shiny(\.min)?\.js["']/.test(body)
}

// Set of x-powered-by values that indicate SSP.
const SSP_POWERED_BY = new Set(["Express", "Shiny Server", "Shiny Server Pro"])

// ---------------------------------------------------------------------------
// Detection
// ---------------------------------------------------------------------------

/**
 * Detect the server type for the given app URL.
 *
 * Detection order follows the original Kotlin implementation:
 *   1. Hostname check for shinyapps.io -> SAI
 *   2. Response header / cookie checks -> SSP or RSC
 *   3. Body check for shiny.js -> SHN
 *   4. Throw if nothing matched
 */
export async function detectServerType(
  appUrl: string,
  httpClient: HttpClient,
): Promise<ServerType> {
  // Step 1: hostname check
  const host = new URL(appUrl).hostname
  if (/^.*\.shinyapps\.io$/.test(host)) {
    return ServerType.SAI
  }

  // Step 2: fetch the app
  const resp = await httpClient.get(appUrl)

  // Step 3a: SSP via x-ssp-xsrf header
  if (resp.headers["x-ssp-xsrf"] !== undefined) {
    return ServerType.SSP
  }

  // Step 3b: SSP via SSP-XSRF cookie in set-cookie header
  const setCookie = resp.headers["set-cookie"] ?? ""
  if (/\bSSP-XSRF\b/.test(setCookie)) {
    return ServerType.SSP
  }

  // Step 3c: SSP via x-powered-by header
  const poweredBy = resp.headers["x-powered-by"]
  if (poweredBy !== undefined && SSP_POWERED_BY.has(poweredBy)) {
    return ServerType.SSP
  }

  // Step 3d: RSC via rscid cookie
  if (/\brscid\b/.test(setCookie)) {
    return ServerType.RSC
  }

  // Step 4: body check for shiny.js
  if (hasShinyJs(resp.body)) {
    return ServerType.SHN
  }

  // Step 5: nothing matched
  throw new Error(
    `Target URL ${appUrl} does not appear to be a Shiny application.`,
  )
}
